### PR TITLE
Add OCR integration feature for NLP-based image and attachment

### PR DESCRIPTION
**Pull request by sudan45**
Addresses
- https://github.com/the-deep/server/issues/1493

## Changes

* LeadPreviewImages change into LeadPreviewAttachment

Mention related users here if any.

## This PR doesn't introduce any:

- [x] temporary files, auto-generated files or secret keys
- [x] n+1 queries
- [x] flake8 issues
- [x] `print`
- [x] typos
- [x] unwanted comments

## This PR contains valid:

- [x] tests
- [x] permission checks (tests here too)
- [ ] translations

### CODE REVIEW

File: apps/lead/models.py [hunk 1]
```diff
@@ -372,14 +372,25 @@ def __str__(self):
...
 
 
-class LeadPreviewImage(models.Model):
+class LeadPreviewAttachment(models.Model):
     """
     NOTE: File can be only used by gallery (when attached to a entry)
     """
+    class AttachementFileType(models.TextChoices):
+        XLSX = 'XLSX', 'XLSX'
+        IMAGE = 'image', 'Image'
```
Comment (thenav56) [inline apps/lead/models.py]: Let's use IntergerChoices.  (Use PositiveSmallIntegerField as field)

File: apps/lead/tests/test_apis.py [hunk 2]
```diff
@@ -1797,7 +1797,16 @@ def test_extractor_callback_url(self, get_file_mock, get_text_mock, index_lead_f
...
+            'images_path': [
+                {
+                    'page_number': 1,
+                    'images': [
+                        'http://random.com/image1.jpeg',
+                        'http://random.com/image1.jpeg'
+                    ],
+                }
+            ],
+            'tables_path': [],
```
Comment (thenav56) [inline apps/lead/tests/test_apis.py]: Is this required?

File: apps/lead/migrations/0050_auto_20240606_0608.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+                ('page_number', models.IntegerField(default=0)),
+                ('type', models.CharField(choices=[('XLSX', 'XLSX'), ('image', 'Image')], max_length=20)),
+                ('file', models.FileField(upload_to='lead-preview/attachments/')),
+                ('file_preview', models.FileField(upload_to='lead-preview/attachments-preview/')),
+                ('lead', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, related_name='images', to='lead.lead')),
+            ],
+        ),
+        migrations.DeleteModel(
+            name='LeadPreviewImage',
+        ),
```
Comment (thenav56) [inline apps/lead/migrations/0050_auto_20240606_0608.py]: We aren't deleting old model, just changing it

File: apps/lead/schema.py [hunk 4]
```diff
@@ -216,6 +217,21 @@ class Meta:
...
 
 
+class LeadPreviewAttachmentsType(DjangoObjectType):
+    file = graphene.Field(FileFieldType)
+    file_preview = graphene.Field(FileFieldType)
+
+    class Meta:
+        model = LeadPreviewAttachment
+        only_fields = (
+            'type',
```
Comment (thenav56) [inline apps/lead/schema.py]: We need to manually define enum class in enums.py and use it for the enum field here

File: apps/lead/schema.py [hunk 5]
```diff
@@ -216,6 +217,21 @@ class Meta:
...
+            'page_number',
+            'order',
+            'file',
+            'file_preview',
```
Comment (thenav56) [inline apps/lead/schema.py]: Not required here

File: apps/lead/schema.py [hunk 6]
```diff
@@ -347,6 +363,7 @@ class Meta:
 
     extraction_status = graphene.Field(LeadExtractionStatusEnum)
     lead_preview = graphene.Field(LeadPreviewType)
+    lead_preview_attachments = graphene.List(graphene.NonNull(LeadPreviewAttachmentsType))
```
Comment (thenav56) [inline apps/lead/schema.py]:
```suggestion
    lead_preview_attachments = graphene.List(graphene.NonNull(LeadPreviewAttachmentsType), required=True)
```

File: apps/lead/schema.py [hunk 7]
```diff
@@ -216,6 +217,21 @@ class Meta:
         )
 
 
+class LeadPreviewAttachmentsType(DjangoObjectType):
```
Comment (thenav56) [inline apps/lead/schema.py]:
```suggestion
class LeadPreviewAttachmentType(DjangoObjectType):
```

File: apps/deepl_integration/serializers.py [hunk 8]
```diff
@@ -63,17 +63,30 @@ class Status(models.IntegerChoices):
     status = serializers.ChoiceField(choices=Status.choices)
 
 
+class ImagePathSerializer(serializers.Serializer):
+    page_number = serializers.IntegerField(required=True)
+    images = serializers.ListField(
+        child=serializers.CharField(allow_blank=True),
+        default=[]
```
Comment (thenav56) [inline apps/deepl_integration/serializers.py]:
```suggestion
        default=[],
```

File: apps/deepl_integration/serializers.py [hunk 9]
```diff
@@ -106,6 +119,7 @@ def create(self, data):
                 lead,
                 data['text_path'],
                 data.get('images_path', [])[:10],   # TODO: Support for more images, too much image will error.
```
Comment (thenav56) [inline apps/deepl_integration/serializers.py:127]: @AdityaKhatri @subinasr Do we will need to change this?

File: apps/deepl_integration/handlers.py [hunk 10]
```diff
@@ -651,18 +652,44 @@ def save_data(
...
-        for image_uri in images_uri:
-            lead_image = LeadPreviewImage(lead=lead)
-            image_obj = RequestHelper(url=image_uri, ignore_error=True).get_file()
-            if image_obj:
-                lead_image.file.save(
-                    os.path.join(image_base_path, os.path.basename(urlparse(image_uri).path)),
-                    image_obj
+
+        attachement_base_path = f'{lead.pk}'
+        images = [dict(item) for item in images_uri]
```
Comment (thenav56) [inline apps/deepl_integration/handlers.py]: Let's do this on serializer, from function typing it should be list of dict

**Review by thenav56 — CHANGES_REQUESTED** [review]
Looks good

File: apps/entry/serializers.py [hunk 11]
```diff
@@ -685,12 +686,21 @@ def validate(self, data):
...
+            if lead_attachment.lead != lead:
                 raise serializers.ValidationError({
-                    'lead_image': f'You don\'t have permission to attach lead image: {lead_image}',
+                    'lead_attachment': f'You don\'t have permission to attach lead attachment: {lead_attachment}',
                 })
-            data['image'] = lead_image.clone_as_deep_file(request.user)
+
+            if lead_attachment.type == LeadPreviewAttachment.AttachementFileType.XLSX:
+                data['entry_attachment'] = EntryAttachment.objects.create(
+                    file=lead_attachment.file,
```
Comment (thenav56) [inline apps/entry/serializers.py]: Let's clone the file as well. If LeadPreviewAttachment is deleted, we need to persist entry attachment data.

File: apps/entry/models.py [hunk 12]
```diff
@@ -21,6 +21,18 @@
 from assisted_tagging.models import DraftEntry
 
 
+class EntryAttachment(models.Model):
+    class EntryFileType(models.IntegerChoices):
```
Comment (thenav56) [inline apps/entry/models.py:28]: Let's add a NOTE on entry will reference it using entry_attachment

File: apps/entry/models.py [hunk 13]
```diff
@@ -21,6 +21,18 @@
...
+        XLSX = 1, 'XLSX'
+
+    entry_file_type = models.PositiveSmallIntegerField(
+        choices=EntryFileType.choices,
+        default=EntryFileType.XLSX
```
Comment (thenav56) [inline apps/entry/models.py]:
```suggestion
        default=EntryFileType.XLSX,
```

File: apps/entry/schema.py [hunk 14]
```diff
@@ -92,7 +101,7 @@ class Meta:
             'lead', 'project', 'analysis_framework', 'information_date', 'order',
             'excerpt', 'dropped_excerpt', 'image', 'tabular_field', 'highlight_hidden',
             'controlled', 'controlled_changed_by',
-            'client_id',
+            'client_id', 'entry_attachment'
```
Comment (thenav56) [inline apps/entry/schema.py]:
```suggestion
            'client_id',
```

File: apps/entry/schema.py [hunk 15]
```diff
@@ -84,6 +85,14 @@ def resolve_geo_selected_options(root, info, **_):
             )
 
 
+class EntryAttachmentType(DjangoObjectType):
+    file = graphene.Field(FileFieldType)
+    file_preview = graphene.Field(FileFieldType)
```
Comment (thenav56) [inline apps/entry/schema.py]:
```suggestion
    file = graphene.Field(FileFieldType, required=True)
    file_preview = graphene.Field(FileFieldType, required=True)
```

File: apps/entry/schema.py [hunk 16]
```diff
@@ -84,6 +85,14 @@ def resolve_geo_selected_options(root, info, **_):
...
+
+    class Meta:
+        model = EntryAttachment
+
```
Comment (thenav56) [inline apps/entry/schema.py:99]: Let's add 
```suggestion
       only_fields = ('id',)
```

File: apps/entry/migrations/0038_auto_20240609_0904.py
Comment (thenav56) [inline apps/entry/migrations/0038_auto_20240609_0904.py:1]: Let's squash all the migrations before merging.

Let's do this at the end
- Revert back all changes
- Delete all files
- Re-create single migration

File: apps/entry/admin.py [hunk 17]
```diff
@@ -86,4 +87,6 @@ class ProjectEntryLabelAdmin(VersionAdmin):
     list_display = ('__str__', 'color')
 
 
+admin.site.register(EntryAttachment)
```
Comment (thenav56) [inline apps/entry/admin.py]: Let's add this as autocomplete in Entry admin model

File: apps/lead/models.py [hunk 18]
```diff
@@ -372,14 +372,25 @@ def __str__(self):
...
+
     lead = models.ForeignKey(
         Lead, related_name='images', on_delete=models.CASCADE,
     )
-    file = models.FileField(upload_to='lead-preview/')
+    order = models.IntegerField(default=0)
+    page_number = models.IntegerField(default=0)
+    type = models.PositiveSmallIntegerField(
+        choices=AttachementFileType.choices,
+        default=AttachementFileType.XLSX
```
Comment (thenav56) [inline apps/lead/models.py]:
```suggestion
        default=AttachementFileType.XLSX,
```

File: apps/lead/models.py [hunk 19]
```diff
@@ -372,14 +372,25 @@ def __str__(self):
...
+    )
+    file = models.FileField(upload_to='lead-preview/attachments/')
+    file_preview = models.FileField(upload_to='lead-preview/attachments-preview/')
 
     def __str__(self):
         return 'Image extracted for {}'.format(self.lead)
```
Comment (thenav56) [inline apps/lead/models.py]:
```suggestion
        return 'Image extracted for {}'.format(self.lead.pk)
```

File: apps/lead/schema.py [hunk 20]
```diff
@@ -216,6 +218,20 @@ class Meta:
         )
 
 
+class LeadPreviewAttachmentType(DjangoObjectType):
+    file = graphene.Field(FileFieldType)
+    file_preview = graphene.Field(FileFieldType)
+    type = graphene.Field(LeadPreviewAttachmentTypeEnum)
```
Comment (thenav56) [inline apps/lead/schema.py]:
```suggestion
    file = graphene.Field(FileFieldType, required=True)
    file_preview = graphene.Field(FileFieldType, required=True)
    type = graphene.Field(LeadPreviewAttachmentTypeEnum, required=True)
```

File: apps/lead/schema.py [hunk 21]
```diff
@@ -347,6 +363,7 @@ class Meta:
 
     extraction_status = graphene.Field(LeadExtractionStatusEnum)
     lead_preview = graphene.Field(LeadPreviewType)
+    lead_preview_attachment = graphene.List(graphene.NonNull(LeadPreviewAttachmentType), required=True)
```
Comment (thenav56) [inline apps/lead/schema.py]:
```suggestion
    lead_preview_attachments = graphene.List(graphene.NonNull(LeadPreviewAttachmentType), required=True)
```

File: apps/deepl_integration/serializers.py [hunk 22]
```diff
@@ -63,17 +63,30 @@ class Status(models.IntegerChoices):
...
     Serialize deepl extractor
     """
     url = serializers.CharField(required=False)
     # Data fields
-    images_path = serializers.ListField(
-        child=serializers.CharField(allow_blank=True),
-        required=False, default=[],
-    )
+    images_path = serializers.ListSerializer(child=ImagePathSerializer(required=False))
+    tables_path = serializers.ListSerializer(child=TablePathSerializer(required=False))
```
Comment (thenav56) [inline apps/deepl_integration/serializers.py]:
```suggestion
    images_path = serializers.ListSerializer(child=ImagePathSerializer(required=True), required=False)
    tables_path = serializers.ListSerializer(child=TablePathSerializer(required=True), required=False)
```

Also, let's check this behavior if user can send Null.

File: apps/deepl_integration/serializers.py [hunk 23]
```diff
@@ -63,17 +63,30 @@ class Status(models.IntegerChoices):
     status = serializers.ChoiceField(choices=Status.choices)
 
 
+class ImagePathSerializer(serializers.Serializer):
+    page_number = serializers.IntegerField(required=True)
+    images = serializers.ListField(
+        child=serializers.CharField(allow_blank=True),
+        default=[],
+    )
```
Comment (thenav56) [inline apps/deepl_integration/serializers.py:71]: Check this behavior, will there be image if there is page_number?

File: apps/entry/schema.py [hunk 24]
```diff
@@ -84,6 +85,14 @@ def resolve_geo_selected_options(root, info, **_):
             )
 
 
+class EntryAttachmentType(DjangoObjectType):
+    file = graphene.Field(FileFieldType)
+    file_preview = graphene.Field(FileFieldType)
+
```
Comment (thenav56) [inline apps/entry/schema.py:93]: Let's add a type for the enum

File: apps/deepl_integration/handlers.py [hunk 25]
```diff
@@ -651,18 +652,42 @@ def save_data(
...
-                lead_image.file.save(
-                    os.path.join(image_base_path, os.path.basename(urlparse(image_uri).path)),
-                    image_obj
+            for image in image_uri['images']:
+                lead_attachement = LeadPreviewAttachment(lead=lead)
+                image_obj = RequestHelper(url=image, ignore_error=True).get_file()
+                if image_obj:
+                    lead_attachement.file.save(
+                        os.path.join(attachement_base_path, os.path.basename(urlparse(image).path)),
+                        image_obj
```
Comment (thenav56) [inline apps/deepl_integration/handlers.py]:
```suggestion
```suggestion
                        os.path.join(
                            attachement_base_path,
                            os.path.basename(urlparse(image).path),
                         ),
                         image_obj,
```
```

File: apps/deepl_integration/handlers.py [hunk 26]
```diff
@@ -651,18 +652,42 @@ def save_data(
...
+                    )
+                    lead_attachement.page_number = image_uri['page_number']
+                    lead_attachement.type = LeadPreviewAttachment.AttachementFileType.IMAGE
+                    lead_attachement.file_preview = lead_attachement.file
+
+                lead_attachement.save()
```
Comment (thenav56) [inline apps/deepl_integration/handlers.py]:
```suggestion
                if image_obj:
                    lead_attachement = LeadPreviewAttachment(lead=lead)
                    ......
                    lead_attachement.save()
```

File: apps/deepl_integration/handlers.py [hunk 27]
```diff
@@ -651,18 +652,42 @@ def save_data(
...
+
+        for table in table_uri:
+            lead_attachement = LeadPreviewAttachment(lead=lead)
```
Comment (thenav56) [inline apps/deepl_integration/handlers.py]:
```suggestion
            if table_img:
                lead_attachement = LeadPreviewAttachment(lead=lead)
```

File: apps/deepl_integration/handlers.py [hunk 28]
```diff
@@ -651,18 +652,42 @@ def save_data(
...
+            table_img = RequestHelper(url=table['image_link'], ignore_error=True).get_file()
+            table_attahcment = RequestHelper(url=table['content_link'], ignore_error=True).get_file()
```
Comment (thenav56) [inline apps/deepl_integration/handlers.py]:
```suggestion
            table attachment = RequestHelper(url=table['content_link'], ignore_error=True).get_file()
```

**Review by thenav56 — CHANGES_REQUESTED** [review]

File: apps/entry/serializers.py [hunk 29]
```diff
@@ -635,6 +634,11 @@ def validate_lead(self, lead):
             raise serializers.ValidationError('Changing lead is not allowed')
         return lead
 
+    def validate_lead_attachment(self, lead_attachment):
+        if int(self.initial_data.get('lead')) != lead_attachment.lead.id:
+            raise serializers.ValidationError("Don't have access to this lead")
+        return lead_attachment
```
Comment (thenav56) [inline apps/entry/serializers.py]: Isolated validations, use validate for complex checks

File: apps/entry/serializers.py [hunk 30]
```diff
@@ -643,7 +647,7 @@ def validate(self, data):
         request = self.context['request']
         image = data.get('image')
```
Comment (thenav56) [inline apps/entry/serializers.py]: We need to handle this as well

File: apps/lead/filter_set.py [hunk 31]
```diff
@@ -552,6 +553,25 @@ def filter_title(self, qs, name, value):
         return qs.filter(title__icontains=value).distinct()
 
 
+class LeadPreviewAttachmentGQFilterSet(UserResourceGqlFilterSet):
+    type = MultipleInputFilter(LeadPreviewAttachmentTypeEnum, field_name='type')
+    exclude_attachment_id = IDListFilter(method='filter_exclude_lead_attachment_id')
```
Comment (thenav56) [inline apps/lead/filter_set.py]:
```suggestion
    exclude_attachment_ids = IDListFilter(method='filter_exclude_lead_attachment_ids')
```

**Review by thenav56 — CHANGES_REQUESTED** [review]

File: apps/deepl_integration/serializers.py (hunk 9 above)
Comment (subinasr) [reply apps/deepl_integration/serializers.py:127]: Safar: For the time being, we can just process 10 items for tables_path as well. Just to make this consistent.
Later, we can get all of these on async fashion.

File: apps/entry/schema.py [hunk 32]
```diff
@@ -103,6 +117,8 @@ class Meta:
     verified_by_count = graphene.Int(required=True)
     review_comments_count = graphene.Int(required=True)
     draft_entry = graphene.ID(source="draft_entry_id")
+    entry_attachment = graphene.Field(EntryAttachmentType, required=False)
+    canonical_preview_image = graphene.String(required=False)
```
Comment (subinasr) [inline apps/entry/schema.py]: We can remove this now

File: apps/entry/serializers.py [hunk 33]
```diff
@@ -641,16 +641,26 @@ def validate(self, data):
...
+        # validate entry tag type
+        entry_type = data.get('entry_type')
+        if entry_type and entry_type == Entry.TagType.ATTACHMENT and lead_attachment is None:
+            raise serializers.ValidationError({
+                'lead_attachment': f'LeadPreviewAttachment is required with entry_tag is {entry_type}'
+            })
+
+        elif entry_type and entry_type == Entry.TagType.EXCERPT and data.get('excerpt') is None:
+            raise serializers.ValidationError({
+                'excerpt': f'EXCERPT is required when entry_tag is {entry_type}'
```
Comment (subinasr) [inline apps/entry/serializers.py:662]: The audience of these error message is the end user.

**Review by susilnem — APPROVED** [review]
LGTM!

File: apps/deepl_integration/handlers.py [hunk 34]
```diff
@@ -683,13 +723,18 @@ def save_lead_data_using_connector_lead(
...
+        # Save extracted images as LeadPreviewAttachment instances
         # TODO: The logic is same for unified_connector leads as well. Maybe have a single func?
-        for connector_lead_preview_image in connector_lead.preview_images.all():
-            lead_image = LeadPreviewImage(lead=lead)
-            lead_image.file.save(
-                connector_lead_preview_image.image.name,
-                connector_lead_preview_image.image,
+        for connector_lead_attachment in connector_lead.preview_images.all():
+            lead_attachment = LeadPreviewAttachment(lead=lead)
+            lead_attachment.order = connector_lead_attachment.order
```
Comment (thenav56) [inline apps/deepl_integration/handlers.py:730]: Missing lead_attachmend.save()

File: apps/unified_connector/migrations/0009_auto_20240618_0924.py [hunk 35]
```diff
@@ -0,0 +1,29 @@
...
+                ('page_number', models.IntegerField(default=0)),
+                ('type', models.PositiveSmallIntegerField(choices=[(1, 'XLSX'), (2, 'Image')], default=1)),
+                ('file', models.FileField(upload_to='connector-lead/attachments/')),
+                ('file_preview', models.FileField(upload_to='connector-lead/attachments-preview/')),
+                ('connector_lead', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, related_name='preview_images', to='unified_connector.connectorlead')),
+            ],
+        ),
+        migrations.DeleteModel(
+            name='ConnectorLeadPreviewImage',
+        ),
```
Comment (thenav56) [inline apps/unified_connector/migrations/0009_auto_20240618_0924.py]: Danger

**Review by thenav56 — CHANGES_REQUESTED** [review]

File: apps/lead/migrations/0051_auto_20240625_0509.py [hunk 36]
```diff
@@ -0,0 +1,45 @@
...
+
+from django.db import migrations, models
+
+
+def set_file_preview(apps, schema_editor):
+    LeadPreviewAttachment = apps.get_model('lead', 'LeadPreviewAttachment')
+    lead_attachments = LeadPreviewAttachment.objects.all()
+    for lead_attachment in lead_attachments:
+        lead_attachment.file_preview = lead_attachment.file
+        lead_attachment.type = 2  # default type is image
```
Comment (thenav56) [inline apps/lead/migrations/0051_auto_20240625_0509.py]: Let's use LeadPreviewAttachment.objects.update(.....)

File: server.diff
Comment (thenav56) [inline server.diff:1]: Clean-up required

File: apps/lead/migrations/0051_auto_20240625_0509.py [hunk 37]
```diff
@@ -0,0 +1,45 @@
...
+        migrations.AddField(
+            model_name='leadpreviewattachment',
+            name='file_preview',
+            field=models.FileField(upload_to='lead-preview/attachments-preview/')
+        ),
+        migrations.AddField(
+            model_name='leadpreviewattachment',
+            name='type',
+            field=models.PositiveSmallIntegerField(choices=[(1, 'XLSX'), (2, 'Image')],max_length=20)
+        ),
```
Comment (thenav56) [inline apps/lead/migrations/0051_auto_20240625_0509.py:35]: This is missing AlterField for file (`upload_to` has been changed)

File: apps/unified_connector/migrations/0010_auto_20240625_0806.py [hunk 38]
```diff
@@ -0,0 +1,44 @@
+# Generated by Django 3.2.25 on 2024-06-25 08:06
+
+from django.db import migrations, models
+
+def set_file_preview(apps, schema_editor):
+    ConnectorLeadPreviewAttachment = apps.get_model('unified_connector', 'ConnectorLeadPreviewAttachment')
+    connector_lead_attachments = ConnectorLeadPreviewAttachment.objects.all()
+    for connector_lead_attachment in connector_lead_attachments:
+        connector_lead_attachment.file_preview = connector_lead_attachment.file
+        connector_lead_attachment.type = 2  # default type is image
```
Comment (thenav56) [inline apps/unified_connector/migrations/0010_auto_20240625_0806.py]: Use ConnectorLeadPreviewAttachment.objects.update(..)

**Review by thenav56 — CHANGES_REQUESTED** [review]
Looks good.

File: apps/entry/migrations/0038_auto_20240609_0904.py
Comment (thenav56) [inline apps/entry/migrations/0038_auto_20240609_0904.py:1]: Delete migrations

File: apps/entry/models.py [hunk 39]
```diff
@@ -19,6 +25,64 @@
...
+        if lead_attachment.type == LeadPreviewAttachment.AttachmentFileType.IMAGE:
+            entry_attachment.file.save(lead_attachment_file_name, lead_attachment.file)
+            entry_attachment.file_preview = entry_attachment.file
+        else:
+            entry_attachment.file.save(lead_attachment_file_name, lead_attachment.file)
+            entry_attachment.file_preview.save(lead_attachment_file_preview_name, lead_attachment.file_preview)
+        entry_attachment.save()
+        return entry_attachment
+
+    def get_file_url(self):
```
Comment (thenav56) [inline apps/entry/models.py]: Let's create a generic function, move the function to gallery/utils.py

```suggestion
    def get_file_url(obj, url):
```

File: apps/gallery/enums.py [hunk 40]
```diff
@@ -0,0 +1,6 @@
+import graphene
+
+
+class ModuleTypeEnum(graphene.Enum):
```
Comment (thenav56) [inline apps/gallery/enums.py]: Let's change the name to PrivateFileModuleType

File: apps/gallery/views.py [hunk 41]
```diff
@@ -82,6 +83,24 @@ def get(self, request, uuid=None, filename=None):
         )
 
 
+class AttachmentFileView(views.APIView):
```
Comment (thenav56) [inline apps/gallery/views.py]:
```suggestion
class PrivateModuleFileView(views.APIView):
```

File: apps/gallery/views.py [hunk 42]
```diff
@@ -82,6 +83,24 @@ def get(self, request, uuid=None, filename=None):
...
+            id = force_text(urlsafe_base64_decode(identifier))
+            qs = get_object_or_404(EntryAttachment, id=id)
+            if qs:
+                return redirect(request.build_absolute_uri(qs.file.url))
+            return response.Response({
+                'error': 'File doesn\'t exists',
+            }, status=status.HTTP_404_NOT_FOUND)
+        return response.Response({
+            'error': 'Access Forbidden, Contact Admin',
+        }, status=status.HTTP_403_FORBIDDEN)
```
Comment (thenav56) [inline apps/gallery/views.py]: Add test as well

```suggestion
        id = force_text(urlsafe_base64_decode(identifier))
        obj = None

        if module == ModuleTypeEnum.ENTRY_ATTACHMENT.value:
            obj = get_object_or_404(EntryAttachment, id=id)
            # Permission check here........
            # EntryAttacment->Entry->Project is member
            # EntryAttacment->Entry->Lead->confidently has access
            # - Check project & visibility permission

        if obj:
            return redirect(request.build_absolute_uri(obj.file.url))
        return response.Response({
            'error': 'Access Forbidden or File does\'t exists, Contact Admin',
        }, status=status.HTTP_403_FORBIDDEN)
```

File: apps/lead/migrations/0051_auto_20240625_0509.py [hunk 43]
```diff
@@ -0,0 +1,45 @@
+# Generated by Django 3.2.25 on 2024-06-25 05:09
+
+from django.db import migrations, models
+
+
+def set_file_preview(apps, schema_editor):
+    LeadPreviewAttachment = apps.get_model('lead', 'LeadPreviewAttachment')
+    LeadPreviewAttachment.objects.update(
+        file_preview=models.F('file'),
+        type=2,
```
Comment (thenav56) [inline apps/lead/migrations/0051_auto_20240625_0509.py]:
```suggestion
        type=2,  # Image
```

File: apps/unified_connector/migrations/0010_auto_20240625_0806.py [hunk 44]
```diff
@@ -0,0 +1,45 @@
+# Generated by Django 3.2.25 on 2024-06-25 08:06
+
+from django.db import migrations, models
+
+
+def set_file_preview(apps, schema_editor):
+    ConnectorLeadPreviewAttachment = apps.get_model('unified_connector', 'ConnectorLeadPreviewAttachment')
+    ConnectorLeadPreviewAttachment.objects.update(
+        file_preview=models.F('file'),
+        type=2,
```
Comment (thenav56) [inline apps/unified_connector/migrations/0010_auto_20240625_0806.py]:
```suggestion
        type=2,  # Image
```

File: apps/entry/schema.py [hunk 45]
```diff
@@ -84,6 +85,19 @@ def resolve_geo_selected_options(root, info, **_):
             )
 
 
+class EntryAttachmentType(DjangoObjectType):
+    lead_attachment_id = graphene.ID(required=True)
```
Comment (thenav56) [inline apps/entry/schema.py]:
```suggestion
    lead_attachment_id = graphene.ID(required=False)
```

File: apps/lead/schema.py [hunk 46]
```diff
@@ -216,6 +219,20 @@ class Meta:
...
+    type = graphene.Field(LeadPreviewAttachmentTypeEnum, required=True)
+
+    class Meta:
+        model = LeadPreviewAttachment
+        only_fields = (
+            'id',
+            'order',
+            'page_number',
+        )
+
```
Comment (thenav56) [inline apps/lead/schema.py:243]: We need this
```suggestion
        @staticmethod
        def get_custom_queryset(queryset, info, **kwargs):
            return get_lead_preview_attachment_qs(info)
```

File: apps/lead/schema.py [hunk 47]
```diff
@@ -480,6 +506,13 @@ class Query:
             page_size_query_param='pageSize'
         )
     )
+    lead_preview_attachments = DjangoPaginatedListObjectField(
+        LeadPreviewAttachmentListType,
+        pagination=PageGraphqlPagination(
+            page_size_query_param='pageSize',
+        )
+    )
+
```
Comment (thenav56) [inline apps/lead/schema.py:528]: We need custom resolver as well

```suggestion
    @staticmethod
    def resolve_lead_preview_attachments(root, info, **kwargs) -> QuerySet:
        return get_lead_preview_attachment_qs(info)
```

**Review by thenav56 — CHANGES_REQUESTED** [review]

File: apps/gallery/utils.py [hunk 48]
```diff
@@ -0,0 +1,20 @@
+from django.conf import settings
+from django.urls import reverse
+from django.utils.encoding import force_bytes
+from django.utils.http import urlsafe_base64_encode
+
+from gallery.enums import PrivateFileModuleType
+
+
+def get_private_file_url(obj, id):
```
Comment (thenav56) [inline apps/gallery/utils.py]: Let's change obj to `private_file_module_type: PrivateFileModuleType`

Call this function without .value, the function should do the .value internally

File: deep/urls.py [hunk 49]
```diff
@@ -435,6 +436,11 @@ def get_api_path(path):
         DeprecatedPrivateFileView.as_view(),
         name='deprecated_gallery_private_url',
     ),
+    path(
+        'external/private-file/<str:module>/<str:identifier>',
```
Comment (thenav56) [inline deep/urls.py]:
```suggestion
        'external/private-file/<str:module>/<str:identifier>/<filename>',
```
> NOTE: filename is only for human consumption

File: apps/gallery/tests/test_apis.py [hunk 50]
```diff
@@ -205,3 +209,38 @@ def save_file_with_api(self, kwargs={}):
...
+
+
+class PrivateAttachmentFileViewTest(TestCase):
+    def setUp(self):
+        # Create a test user
+        self.user = User.objects.create_user(username='testuser', password='testpassword')
+
+        # Create a test entry attachment
+        self.project = Project.objects.create()
+        self.lead = Lead.objects.create(project=Project)
```
Comment (thenav56) [inline apps/gallery/tests/test_apis.py]: 3 types of leads by confidentiality

File: apps/gallery/tests/test_apis.py [hunk 51]
```diff
@@ -205,3 +209,38 @@ def save_file_with_api(self, kwargs={}):
...
+        )
+        self.url = reverse('private_attachment_file_view', kwargs={
+            'module': PrivateFileModuleType.ENTRY_ATTACHMENT.value,
+            'identifier': urlsafe_base64_encode(force_bytes(self.attachment.id)),
+        })
+
+    def test_access_by_authenticated_user(self):
+        self.authenticate()
+        response = self.client.get(self.url)
+        self.assert_200(response)
```
Comment (thenav56) [inline apps/gallery/tests/test_apis.py]: this should be 403

File: apps/gallery/tests/test_apis.py [hunk 52]
```diff
@@ -205,3 +209,38 @@ def save_file_with_api(self, kwargs={}):
         return response.data['id']
 
     # NOTE: Test for files
+
+
+class PrivateAttachmentFileViewTest(TestCase):
+    def setUp(self):
+        # Create a test user
+        self.user = User.objects.create_user(username='testuser', password='testpassword')
```
Comment (thenav56) [inline apps/gallery/tests/test_apis.py]: 3 types of users.
- Normal
- Member (Non-confidentiality)
- Member (confidentiality)

File: apps/gallery/tests/test_apis.py [hunk 53]
```diff
@@ -205,3 +209,38 @@ def save_file_with_api(self, kwargs={}):
...
+            lead=self.lead,
+            project=self.project,
+            entry=self.attachment
+        )
+        self.url = reverse('private_attachment_file_view', kwargs={
+            'module': PrivateFileModuleType.ENTRY_ATTACHMENT.value,
+            'identifier': urlsafe_base64_encode(force_bytes(self.attachment.id)),
+        })
+
+    def test_access_by_authenticated_user(self):
```
Comment (thenav56) [inline apps/gallery/tests/test_apis.py]: To check
- Non-logged in -> 403?
- Normal User -> 403
- Member (N-C) -> 403
- Member (C) -> 200

For id > 1000000 -> all should be 403

File: apps/entry/models.py (hunk 39 above)
Comment (thenav56) [reply apps/entry/models.py]: Clean-up

File: apps/gallery/views.py [hunk 54]
```diff
@@ -82,6 +83,23 @@ def get(self, request, uuid=None, filename=None):
...
+class PrivateAttachmentFileView(views.APIView):
+    permission_classes = [permissions.IsAuthenticated]
+
+    def get(self, request, module=None, identifier=None):
+        id = force_text(urlsafe_base64_decode(identifier))
+        user = request.user
+        obj = None
+        if module == PrivateFileModuleType.ENTRY_ATTACHMENT.value:
+            obj = get_object_or_404(EntryAttachment, id=id)
+            obj.entry.get_for(user)
```
Comment (thenav56) [inline apps/gallery/views.py]: Need to make sure this work.

**Review by thenav56 — CHANGES_REQUESTED** [review]

File: apps/gallery/tests/test_apis.py [hunk 55]
```diff
@@ -205,3 +213,122 @@ def save_file_with_api(self, kwargs={}):
...
+        self.force_login(self.normal_user)
+        response = self.client.get(self.url2)
+        self.assertEqual(403, response.status_code)
+
+    def test_access_by_non_member_user(self):
+        self.force_login(self.member_user)
+        response = self.client.get(self.url4)
+        self.assertEqual(403, response.status_code)
+
+    def test_access_by_member_user(self):
```
Comment (thenav56) [inline apps/gallery/tests/test_apis.py:335]: Missing member with non-confidential only permisison

File: apps/gallery/utils.py [hunk 56]
```diff
@@ -0,0 +1,33 @@
...
+from django.urls import reverse
+from django.utils.encoding import force_bytes
+from django.utils.http import urlsafe_base64_encode
+
+from gallery.enums import PrivateFileModuleType
+from deep.permissions import ProjectPermissions as PP
+from lead.models import Lead
+
+
+def get_private_file_url(private_file_module_type: PrivateFileModuleType, entry):
```
Comment (thenav56) [inline apps/gallery/utils.py]: This is a generic function, we shouldn't pass entry here

```suggestion
def get_private_file_url(private_file_module_type: PrivateFileModuleType, id: int, filename: str):
```

File: apps/gallery/utils.py [hunk 57]
```diff
@@ -0,0 +1,33 @@
...
+            kwargs={
+                'module': private_file_module_type.value,
+                'identifier': urlsafe_base64_encode(force_bytes(entry.id)),
+                'filename': entry.entry_attachment.file.name
+            }
+        )
+    )
+
+
+def check_private_entry_qs(user, project, level):
```
Comment (thenav56) [inline apps/gallery/utils.py]: We need to update the name of this function.

With something like this
```suggestion
def check_entry_confidential_level_permisison(user, project, level):
```

**Review by thenav56 — CHANGES_REQUESTED** [review]

**Review by thenav56 — APPROVED** [review]

**Review by susilnem — APPROVED** [review]
GTM!

**Pull request merged**